src/systems/acoustic_comms/AcousticComms.cc

```diff
@@ -0,0 +1,186 @@
+
+using namespace gz;
+using namespace sim;
+using namespace systems;
+
+class AcousticComms::Implementation
+{
+  // Default max range for acoustic comms in metres.
+  public: double maxRange = 1000.0;
+
+  // Default speed of sound in air (m/s).
+  public: double speedOfSound = 343.0;
+
+  public: double DistanceBetweenBodies(
+              math::Vector3<double> _src,
+              math::Vector3<double> _dst);
+};
+
+double AcousticComms::Implementation::DistanceBetweenBodies(
+    math::Vector3<double> _src, math::Vector3<double> _dst)
+{
+  double x = _src.X() - _dst.X();
+  double y = _src.Y() - _dst.Y();
+  double z = _src.Z() - _dst.Z();
+  return sqrt(pow(x, 2) + pow(y, 2) + pow(z, 2));
+}
+
+AcousticComms::AcousticComms()
+  : dataPtr(gz::utils::MakeUniqueImpl<Implementation>())
+{
+}
+
+void AcousticComms::Load(
+    const Entity &_entity,
+    std::shared_ptr<const sdf::Element> _sdf,
+    EntityComponentManager &_ecm,
+    EventManager &_eventMgr)
+{
+  if (_sdf->HasElement("max_range"))
+  {
+    this->dataPtr->maxRange = _sdf->Get<double>("max_range");
+  }
+  if (_sdf->HasElement("speed_of_sound"))
+  {
+    this->dataPtr->speedOfSound = _sdf->Get<double>("speed_of_sound");
+  }
+}
+
+void AcousticComms::Step(
+    const UpdateInfo &_info,
+    const comms::Registry &_currentRegistry,
+    comms::Registry &_newRegistry,
+    EntityComponentManager &_ecm)
+{
+  // Initialize entity if needed.
+  for (auto & [address, content] : _currentRegistry)
+  {
+    if (content.entity == kNullEntity)
+    {
+      auto entities = entitiesFromScopedName(content.modelName, _ecm);
+      if (entities.empty())
+        continue;
+
+      auto entityId = *(entities.begin());
+      if (entityId == kNullEntity)
+        continue;
+
+      _newRegistry[address].entity = entityId;
+    }
+  }
+
+  for (auto & [address, content] : _currentRegistry)
+  {
+    // Reference to the outbound queue for this address.
+    auto &outbound = content.outboundMsgs;
+
+    // Is the source address bound?
+    auto itSrc = _currentRegistry.find(address);
+    bool srcAddressBound = itSrc != _currentRegistry.end();
+
+    // Is the source address attached to a model?
+    bool srcAddressAttachedToModel =
+      srcAddressBound && itSrc->second.entity != kNullEntity;
+
+    comms::DataQueue newOutbound;
+
+    if (srcAddressAttachedToModel)
+    {
+      // All these messages need to be processed.
+      for (auto &msg : outbound)
+      {
+        // Is the destination address bound?
+        auto itDst = _currentRegistry.find(msg->dst_address());
+        bool dstAddressBound = itDst != _currentRegistry.end();
+
+        // Is the destination address attached to a model?
+        bool dstAddressAttachedToModel =
+          dstAddressBound && itDst->second.entity != kNullEntity;
+
+        if (!dstAddressAttachedToModel)
+          continue;
+
+        // Calculate distance between the bodies.
+        auto poseSrc = worldPose(itSrc->second.entity, _ecm).Pos();
```

Comment:
@adityapande-1995 like you've pointed out elsewhere, we need the position of the sender at the time the message was emitted to model attenuation. Generally speaking, different communication models may need different bits of information  to be sent with the messages. I'm not sure what @caguero had in mind, but I'd think we need to compose messages. Type erasure would be handy, but templating the comms infrastructure would do too.

Reply:
Checks for ``transmitter_position`` in header of the dataframe msg : https://github.com/gazebosim/gz-sim/pull/1608/commits/aa779fa203d3613e72ef305848ba157c1aa7e53f